src/next/main.ts

```diff
@@ -0,0 +1,143 @@
+import * as fse from 'fs-extra';
+import * as uuid from "uuid";
+import TelemetryReporter from "vscode-extension-telemetry";
+import { ErrorCode } from './ErrorCode';
+import { EventName } from "./EventName";
+
+let _isDebug: boolean = false;
+let reporter: TelemetryReporter;
+
+/**
+ * Initialize TelemetryReporter by parsing attributes from a JSON file.
+ * It reads these attributes: publisher, name, version, aiKey.
+ * @param jsonFilepath absolute path of a JSON file.
+ */
+export async function initilizeFromJsonFile(jsonFilepath: string, _debug?: boolean): Promise<void> {
+    if (await fse.pathExists(jsonFilepath)) {
+        const { publisher, name, version, aiKey } = await fse.readJSON(jsonFilepath);
+        initilizeFromAttributes(`${publisher}.${name}`, version, aiKey, !!_debug);
+    } else {
+        throw new Error(`The Json file '${jsonFilepath}' does not exist.`);
+    }
+}
+
+/**
+ * Initialize TelemetryReporter from given attributes.
+ * @param extensionId Idenfier of the extension, used as prefix of EventName in telemetry data.
+ * @param version Version of the extension.
+ * @param aiKey Key of Application Insights.
+ */
+export function initilizeFromAttributes(extensionId: string, version: string, aiKey: string, _debug?: boolean): void {
+    if (reporter) {
+        throw new Error("TelemetryReporter already initilized.");
+    }
+    if (aiKey) {
+        reporter = new TelemetryReporter(extensionId, version, aiKey);
+    }
+    _isDebug = !!_debug;
+}
+
+/**
+ * Wrap callback for a command to auto send OPEARTION_START, OPERATION_END, ERROR telemetry.
+ * @param commandName 
+ * @param cb The callback function for the command.
+ * @returns A new callback with telemetry logic injected.
+ */
+export function getWrappedCommand(commandName: string, cb: (...args: any[]) => any): (...args: any[]) => any {
+    return async (...args: any[]) => {
+        let error = undefined;
+        let oId = uuid.v4();
+        const startAt: number = Date.now();
+        try {
+            sendOpStart(oId, commandName);
+            return await cb(...args);
+        } catch (e) {
+            error = e;
+            sendError(oId, error);
+        } finally {
+            const duration = Date.now() - startAt;
+            sendOpEnd(oId, commandName, duration, error);
+        }
+    }
+};
+
+/**
+ * Send ERROR telemetry data, supposed to be used when an error occurs.
+ * @param oId Operation Id.
+ * @param errorObject An object containing error details.
+ */
+export function sendError(oId: string, errorObject: any) {
+    if (!errorObject || errorObject.errorCode === ErrorCode.NO_ERROR) {
+        if (_isDebug) {
+            console.warn("No error indicated by: ", errorObject);
+        }
+    } else {
+        const errorProperties: Properties = getErrorProperties(errorObject);
+        report(EventName.ERROR, { oId, ...errorProperties });

```

Comment:
I would separate the cases of whether `oId` exists.
```ts
if (oId) {
  report(..., {oId, ...errorProperties});
  return;
}

report(..., errorProperties};
```
In this way, there will not be an entry named `oId` when it's not available.